### ISSUE TITLE
I do not quite get the difference between sirens and Shadowenv

### ISSUE THREAD

**Issue by hauleth** [Shopify/shadowenv#56]
Sorry for asking there, but I couldn't find any other discussion channel.

---

In `README` there is fragment:

> Unlike other tools like `direnv`, this has the interesting property of allowing us to do things like simulate `chruby` reset upon entry into a directory without the user having `chruby` installed (and undo these changes to the environment when `cd`'ing back out)

But it is still possible in `direnv` via `layout` function. So is there anything more except different configuration language (sh vs Ketos)?

**Comment by burke** [comment]
Yup, direnv is basically capable of everything Shadowenv is capable of; the main difference is that shadowenv can't do anything other than set environment variables with a bit of conditional logic, and has strict execution resource restrictions, so I feel comfortable running it in shell hooks and other synchronous operations. It can never run for longer than (IIRC) 100ms.

**Comment by fzakaria** [comment]
It would be neat to offer some precanned shadowlips files or examples similar to direnv (https://direnv.net/docs/ruby.html)
It's not clear from the docs how to use shadowlisp to maybe accomplish some simple tasks.

At shopify, is it used just to purge the current environment and then rely on Nix to actually choose the Ruby version?
(nix-shell)

**Comment by lugray** [comment]
> It would be neat to offer some precanned shadowlips files or examples similar to direnv (https://direnv.net/docs/ruby.html)
It's not clear from the docs how to use shadowlisp to maybe accomplish some simple tasks.

There's a ruby example in the README - what are you seeking to accomplish that that fails to give a launching point for?

> At shopify, is it used just to purge the current environment and then rely on Nix to actually choose the Ruby version?
(nix-shell)

No, we don't use nix shell.  We make full use of shadowenv to add the right nix paths to the environment for whichever standard shell (bash, zsh, or fish) the user prefers.

**Comment by fzakaria** [comment]
The example in the README only removes environment variables for Ruby.

What's setting it up for the correct version then ?

**Comment by lugray** [comment]
> The example in the README only removes environment variables for Ruby.

You are correct - I scanned that too quickly and missed that it was entirely destructive.  I've added #74.

**Comment by lugray** [comment]
Because you mentioned nix, @fzakaria, a Shadowlisp for ruby from a project using it:
```scheme
(provide "ruby" "2.7.1")

(let ((dev-env "/nix/store/9b7ff3xkcdmqgszxg306z5rvysxn5g4q-shadowenv-ruby-env-shopify"))
  (do
    (env/set "DEV_PROFILE" dev-env)
    (env/prepend-to-pathlist "PATH"    (path-concat dev-env "bin"))
    (env/prepend-to-pathlist "PKG_CONFIG_PATH" (path-concat dev-env "lib" "pkgconfig"))
    (env/prepend-to-pathlist "MANPATH" (path-concat dev-env "share" "man"))))

(when-let ((ruby-root (env/get "RUBY_ROOT")))
 (env/remove-from-pathlist "PATH" (path-concat ruby-root "bin"))
 (when-let ((gem-root (env/get "GEM_ROOT")))
   (env/remove-from-pathlist "PATH" (path-concat gem-root "bin")))
 (when-let ((gem-home (env/get "GEM_HOME")))
   (env/remove-from-pathlist "PATH" (path-concat gem-home "bin"))))
(env/set "GEM_PATH" ())
(env/set "GEM_HOME" ())
(env/set "RUBYOPT" ())
(env/set "RUBYLIB" ())

; generated from /nix/store/pm9vzwm9n5xp6zbxaaz6h1pd2fggp0nc-ruby-2.7.1/nix-support/setup-hook
(env/prepend-to-pathlist "GEM_PATH" "/nix/store/pm9vzwm9n5xp6zbxaaz6h1pd2fggp0nc-ruby-2.7.1/lib/ruby/gems/2.7.0")
(env/prepend-to-pathlist "RUBYLIB" "/nix/store/pm9vzwm9n5xp6zbxaaz6h1pd2fggp0nc-ruby-2.7.1/lib/ruby/site_ruby")
(env/prepend-to-pathlist "RUBYLIB" "/nix/store/pm9vzwm9n5xp6zbxaaz6h1pd2fggp0nc-ruby-2.7.1/lib/ruby/site_ruby/2.7.0")
(env/prepend-to-pathlist "RUBYLIB" "/nix/store/pm9vzwm9n5xp6zbxaaz6h1pd2fggp0nc-ruby-2.7.1/lib/ruby/site_ruby/2.7.0/x86_64-darwin")

; generated from /nix/store/gvywl8dyps5qiinybby6jqp2lwkkxdc3-bundler-2.1.4/nix-support/setup-hook
(env/prepend-to-pathlist "GEM_PATH" "/nix/store/gvywl8dyps5qiinybby6jqp2lwkkxdc3-bundler-2.1.4/lib/ruby/gems/2.7.0")
(env/prepend-to-pathlist "RUBYLIB" "/nix/store/gvywl8dyps5qiinybby6jqp2lwkkxdc3-bundler-2.1.4/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib")

; generated from /nix/store/d89clwn24y83nzkksrjbhz4l2j9q71zn-gem-bundle-shopify/nix-support/setup-hook
(env/prepend-to-pathlist "GEM_PATH" "/nix/store/d89clwn24y83nzkksrjbhz4l2j9q71zn-gem-bundle-shopify/lib/ruby/gems/2.7.0")
(env/set "BUNDLE_PATH" "/nix/store/d89clwn24y83nzkksrjbhz4l2j9q71zn-gem-bundle-shopify/lib/ruby/gems/2.7.0")
(env/set "RUBYOPT" "-r /nix/store/6yl04bbvfwckvbl3rmnjzp1zzxlr0dk5-nix_store_safeguard.rb")

; some extra ruby activation niceties
(env/set "RUBY_ROOT" "/nix/store/pm9vzwm9n5xp6zbxaaz6h1pd2fggp0nc-ruby-2.7.1")
(env/set "RUBY_ENGINE" "ruby")
(env/set "RUBY_VERSION" "2.7.1")
(env/set "GEM_ROOT" "/nix/store/pm9vzwm9n5xp6zbxaaz6h1pd2fggp0nc-ruby-2.7.1/lib/ruby/gems/2.7.0")
(let ((gem-home "/Users/lisaugray/src/github.com/Shopify/shopify/.dev/gems/2.7.0"))
  (do
    (env/set "GEM_HOME" gem-home)
    (env/prepend-to-pathlist "GEM_PATH" gem-home)
    (env/prepend-to-pathlist "PATH" (path-concat gem-home "bin"))))
```

**Comment by fzakaria** [comment]
@lugray thanks -- I know my comment was a bit curt; I was responding via phone + email.

Thank you for the pull request;
I highly recommend maybe a _recipes_ folder with some nice examples: ruby, nix+ruby, python etc.. :)

I know the documentation says that "tooling should generate" but for many simple projects I don't see why the files cannot be copied + committed into the repository.

I do see why potentially for nix since the hash may change.
What's setting the dependency on the nixpkgs though so they are within the store + not garbage collected ?

**Comment by lugray** [comment]
> What's setting the dependency on the nixpkgs though so they are within the store + not garbage collected ?

That would be the tooling that's generating the Shadowlisp. For us, that's `dev`, an internal tool that sets up project dependencies and gets it ready to run. It sets up a gcroot via the project directory, so that it becomes broken if the project is deleted.

**Comment by donk-shopify** [comment]
No activity in a year
